contributor-advice.md

```diff
@@ -0,0 +1,107 @@
+`JavaConverters` where needed
+* Use the `akka.japi.Pair` class to return tuples
+* Use `scala.compat.java8.FutureConverters` to translate Futures to `CompletionStage`s
+
+
+## Implementation details
+
+### External Dependencies
+
+All the external runtime dependencies for the project, including transitive dependencies, must have an open source license that is equal to, or compatible with, [Apache 2](http://www.apache.org/licenses/LICENSE-2.0).
+
+This must be ensured by manually verifying the license for all the dependencies for the project:
+
+1. Whenever a committer to the project changes a version of a dependency (including Scala) in the build file.
+2. Whenever a committer to the project adds a new dependency.
+3. Whenever a new release is cut (public or private for a customer).
+
+Every external dependency listed in the build file must have a trailing comment with the license name of the dependency.
+
+Which licenses are compatible with Apache 2 are defined in [this doc](http://www.apache.org/legal/3party.html#category-a), where you can see that the licenses that are listed under ``Category A`` automatically compatible with Apache 2, while the ones listed under ``Category B`` needs additional action:
+
+> Each license in this category requires some degree of [reciprocity](http://www.apache.org/legal/3party.html#define-reciprocal); therefore, additional action must be taken in order to minimize the chance that a user of an Apache product will create a derivative work of a reciprocally-licensed portion of an Apache product without being aware of the applicable requirements.
+
+
+### Packages & Scoping
+
+Use `private`, `private[connector]` and `final` extensively to limit the API surface.
+
+| Package                                  | Purpose
+| -----------------------------------------|------------------------
+| `akka.stream.alpakka.connector.javadsl`  | Java-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector.scaladsl` | Scala-only part of the API, normally factories for Sources, Flows and Sinks
+| `akka.stream.alpakka.connector`          | Shared API, eg. settings classes, implementation if thorowly made private
+| `akka.stream.alpakka.connector.impl`     | Optional, implementation in separate package
+
+### Graph stage checklist
+
+* Keep mutable state within the `GraphStageLogic` only 
+* Open connections in `preStart`
+* Release resources in `postStop`
+* Fail early on configuration errors
+* No Blocking At Any Time 
+* Make sure the code is thread-safe
+
+### Keep the code DRY
+
+Avoid duplication of code between different Sources, Sinks and Flows. Extract the common logic to a common abstract
+base `GraphStageLogic` that is inherited by the `GraphStage`s.
+
+In some cases it may be useful to implement the Sources or the Sinks by means of Flows:
+* Source: `Source.maybe.viaMat(MyFlow)(Keep.right)`
+* Sink: `MyFlow.toMat(Sink.ignore)(Keep.right)`
```

Comment:
I'd reword a bit: Sometimes it may be useful for users to provide them with Sink or Source versions of an operation, even if the main concept is implemented (and most useful) as a Flow. This can be easily implemented as (reusing the core Flow implementation):